pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,318 @@
+				return false, retry.Unrecoverable(fmt.Errorf("getting machine, %w", err))
+			}
+			waitErrs[i] = fmt.Errorf("getting node claim, %w", err)
+			continue
+		}
+		// We emitted this event when Deprovisioning was blocked on launching/termination.
+		// This does not block other forms of deprovisioning, but we should still emit this.
+		q.recorder.Publish(deprovisioningevents.Launching(nodeClaim, cmd.Reason))
+		initializedStatus := nodeClaim.StatusConditions().GetCondition(v1beta1.Initialized)
+		if !initializedStatus.IsTrue() {
+			q.recorder.Publish(deprovisioningevents.WaitingOnReadiness(nodeClaim))
+			waitErrs[i] = fmt.Errorf("getting node claim, %w", err)
+			continue
+		}
+		key.Initialized = true
+		// Subtract the last initialization time from the time the command was added to get initialization duration.
+		initLength := initializedStatus.LastTransitionTime.Inner.Time.Sub(cmd.TimeAdded).Seconds()
+		deprovisioningReplacementNodeInitializedHistogram.Observe(initLength)
+	}
+	// If we have any errors, don't continue
+	if err := multierr.Combine(waitErrs...); err != nil {
+		return true, err
+	}
+
+	// All replacements have been provisioned.
+	// All we need to do now is get a successful delete call for each node claim,
+	// then the termination controller will handle the eventual deletion of the nodes.
+	var multiErr error
+	for i := range cmd.Candidates {
+		candidate := cmd.Candidates[i]
+		q.recorder.Publish(deprovisioningevents.Terminating(candidate.Node, candidate.NodeClaim, cmd.Reason)...)
+		if err := nodeclaim.Delete(ctx, q.kubeClient, candidate.NodeClaim); err != nil {
+			if !errors.IsNotFound(err) {
+				multiErr = multierr.Append(multiErr, err)
+			}
+		} else {
+			nodeclaim.TerminatedCounter(cmd.Candidates[i].NodeClaim, cmd.Reason).Inc()
+		}
+	}
+	// If there were any deletion failures, we should requeue.
+	// In the case where we requeue, but the timeout for the command is reached, we'll mark this as a failure.
+	if multiErr != nil {
+		return false, fmt.Errorf("terminating nodeclaims, %w", multiErr)
+	}
+	return false, nil
+}
+
+// Remove fully clears the queue of all references of a hash/command
+func (q *Queue) Remove(cmd *Command) {
+	// Remove all candidates linked to the command
+	for _, candidate := range cmd.Candidates {
+		delete(q.candidateProviderIDToCommand, candidate.ProviderID())
+	}
+	q.RateLimitingInterface.Forget(cmd)
+	q.RateLimitingInterface.Done(cmd)
+}
+
+// CanAdd is a quick check to see if the candidate is already part of a deprovisioning action
+func (q *Queue) CanAdd(ids ...string) error {
+	var err error
+	for _, id := range ids {
+		if _, ok := q.candidateProviderIDToCommand[id]; ok {
+			err = multierr.Append(err, fmt.Errorf("candidate is part of active deprovisioning decision"))
+		}
+	}
+	return err
+}
+
+func (q *Queue) setNodesUnschedulable(ctx context.Context, isUnschedulable bool, candidates ...string) error {
```

Comment:
I'm cautious of using node.spec.Unschedulable, as it may conflict with customers who have manually cordoned nodes. We probably want to simply not touch a node that has been manually cordoned.

Reply:
Totally agree, I had included this here https://github.com/aws/karpenter-core/pull/508, but wanted to wait until we had more information about the taints/api group naming before I changed it. IMO these are independent.